src/optimizer/analyze.cpp

```diff
@@ -0,0 +1,144 @@
+                                                     catalog::col_oid_t columnID, storage::SqlTable *sql_table,
+                                                     terrier::transaction::TransactionContext *scan_txn,
+                                                     unsigned mostFrequentNumber) {
+  // Stats collecting temporary data-structures
+  std::unordered_map<T, unsigned> counting_map;
+  unsigned num_null_values = 0;
+  unsigned num_rows = 0;
+  unsigned top_k = mostFrequentNumber;
+  std::vector<double> histogram_bounds;
+  histogram_bounds.push_back(DBL_MAX);
+  histogram_bounds.push_back(DBL_MIN);
+
+  std::vector<catalog::col_oid_t> current_col_oid_vector{columnID};
+  storage::ProjectionMap projection_list_indices = sql_table->ProjectionMapForOids(current_col_oid_vector);
+  auto initializer =
+      sql_table->InitializerForProjectedColumns(current_col_oid_vector, common::Constants::K_DEFAULT_VECTOR_SIZE);
+  auto *buffer = common::AllocationUtil::AllocateAligned(initializer.ProjectedColumnsSize());
+  storage::ProjectedColumns *projected_column = initializer.Initialize(buffer);
+
+  // scan the results
+  auto iterator_table = sql_table->begin();
+  while (iterator_table != sql_table->end()) {
+    sql_table->Scan(scan_txn, &iterator_table, projected_column);
+    common::RawBitmap *null_bit_map = projected_column->ColumnNullBitmap(projection_list_indices[columnID]);
+    T *col_pointer = reinterpret_cast<T *>(projected_column->ColumnStart(projection_list_indices[columnID]));
+    for (uint32_t i = 0; i < projected_column->NumTuples(); i++) {
+      if (null_bit_map->Test(i)) {
+        if (counting_map.find(col_pointer[i]) != counting_map.end())
+          counting_map[col_pointer[i]]++;
+        else
+          counting_map[col_pointer[i]] = 1;
+
+        histogram_bounds[0] = std::min(static_cast<double>(col_pointer[i]), histogram_bounds[0]);
+        histogram_bounds[1] = std::max(static_cast<double>(col_pointer[i]), histogram_bounds[1]);
+      } else {
+        num_null_values += 1;
+      }
+      num_rows += 1;
+    }
+  }
+
+  std::vector<std::pair<double, double>> intermediate_hist;
+  intermediate_hist.reserve(counting_map.size());
+  for (auto it : counting_map) {
+    intermediate_hist.emplace_back(std::make_pair(static_cast<double>(it.second), static_cast<double>(it.first)));
+  }
+
+  std::sort(intermediate_hist.begin(), intermediate_hist.end());
+  std::reverse(intermediate_hist.begin(), intermediate_hist.end());
+
+  std::vector<double> most_common_values;
+  std::vector<double> most_common_frequencies;
+
+  unsigned index = 0;
+  while (top_k--) {
+    most_common_values.push_back(intermediate_hist[index].second);
+    most_common_frequencies.push_back(intermediate_hist[index].first);
+    index++;
+  }
+
+  terrier::optimizer::ColumnStats col_stats(database_id, table_id, columnID, num_rows, counting_map.size(),
```

Comment:
Since this object has a lot of input arguments, switch to the builder programming pattern for the constructor.